semantic/src/main/kotlin/tools/samt/semantic/SemanticModel.kt

```diff
@@ -0,0 +1,284 @@
+                check(filePackage is PackageType)
+                putAll(filePackage.sourcePackage.types)
+
+                // Add all imports to scope
+                file.imports.forEach { import ->
+                    when (import) {
+                        is TypeImportNode -> {
+                            // Just import one type
+                            val type = global.resolveType(import.name)
+                            if (type != null) {
+                                val name = if (import.alias != null) {
+                                    import.alias!!.name
+                                } else {
+                                    import.name.components.last().name
+                                }
+
+                                putIfAbsent(name, type)
+                            } else {
+                                controller.createContext(file.sourceFile).error {
+                                    message("Import '${import.name}' could not be resolved")
+                                    highlight("unresolved import", import.location)
+                                }
+                            }
+                        }
+
+                        is WildcardImportNode -> {
+                            // Import all types from the package
+                            val type = global.resolveType(import.name)
+                            if (type != null) {
+                                if (type is PackageType) {
+                                    type.sourcePackage.types.forEach { (name, type) ->
+                                        putIfAbsent(name, type)
+                                    }
+                                } else {
+                                    controller.createContext(file.sourceFile).error {
+                                        message("Import '${import.name}.*' must point to a package and not a type")
+                                        highlight(
+                                            "illegal wildcard import", import.location, suggestChange = "import ${
+                                                import.name.components.dropLast(1).joinToString(".") { it.name }
+                                            }.*")
+                                    }
+                                }
+                            } else {
+                                controller.createContext(file.sourceFile).error {
+                                    message("Import '${import.name}.*' could not be resolved")
+                                    highlight("unresolved import", import.location)
+                                }
+                            }
+                        }
+                    }
+                }
+
+                // Add all top-level packages to scope
+                topLevelPackages.forEach { putIfAbsent(it.packageName, it) }
+
+                // Add built-in types
```

Comment:
Should we register those in the global scope and prevent any new types from shadowing them? Or at least give out a warning if you're shadowing any built-in types.

Reply:
I looked at what Kotlin does and the don't stop you for defining a type called `Int` or `String` in any way. My current opproach creates a flat type lookup scope for every file, this last part representing the "global" scope if you want.

I added a warning to inform the user when a pre-defined type name is shadowed